# Thống kê mô tả trên tập dữ liệu về phân loại chất lượng rượu đỏ

## 1. Chuẩn bị dữ liệu

In [ ]:
import pandas as pd 
wine_data = pd.read_csv("winequality-red.csv")
wine_data = wine_data[['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality']]
display(wine_data.head())


## 2. Loại bỏ dữ liệu trùng lặp

Trước hết, chúng ta cần kiểm tra xem có dữ liệu trùng lặp hay không

In [9]:
duplicate_count = wine_data.duplicated().sum()
duplicate_count


240

Sau khi kiểm tra, ta nhận thấy có rất nhiều dòng dữ liệu trùng lặp, nên chúng ta sẽ thực hiện loại bỏ (ở đây có 240 dòng dữ liệu trùng lặp)

In [10]:
wine_data.head()
wine_data = wine_data.drop_duplicates()


## 3. Thay thế dữ liệu thiếu, thay đổi định dạng của dữ liệu

Trước hết, cần kiểm tra định dạng của bộ dữ liệu 

In [ ]:
wine_data_info = wine_data.dtypes
wine_data_info


fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

Sau khi kiểm tra, ta nhận thấy các định dạng đều đồng nhất, và không có vấn đề gì, nên ta bỏ qua bước này. Tiếp theo chúng ta sẽ kiểm tra xem bộ dữ liệu này có bị thiếu hay không (kiêm tra xem bộ nào không có dữ liệu từng cột, hoặc dữ liệu = 0)

In [14]:
missing_info = wine_data.isnull().sum()
missing_info


fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

Dựa vào kết quả kiểm tra này, ta kết luận bộ dữ liệu này đã đồng nhất về định dạng, và không có dữ liệu bị thiếu.

## 4. Tính toán và thống kê mô tả 

Tới đây, ta sẽ thống kê mô tả bộ dữ liệu này.

In [ ]:
import numpy as np
from scipy import stats

# Tính toán các thống kê mô tả cho từng cột số liệu
desc_stats = {}

for col in wine_data.columns:
    if wine_data[col].dtype in [np.float64, np.int64]:
        data = wine_data[col].values
        
        mean = np.mean(data)
        median = np.median(data)
        mode = stats.mode(data, keepdims=True)[0][0]
        var = np.var(data, ddof=1)
        std = np.std(data, ddof=1)
        data_range = np.ptp(data)
        percentile_25 = np.percentile(data, 25)
        percentile_50 = np.percentile(data, 50)
        percentile_75 = np.percentile(data, 75)
        iqr = stats.iqr(data)
        
        desc_stats[col] = {
            'mean': mean,
            'median': median,
            'mode': mode,
            'variance': var,
            'std_dev': std,
            'range': data_range,
            'percentile_25': percentile_25,
            'percentile_50': percentile_50,
            'percentile_75': percentile_75,
            'iqr': iqr
        }

desc_stats_df = pd.DataFrame(desc_stats).T
desc_stats_df


mean   median     mode     variance    std_dev  \
fixed acidity          8.310596   7.9000   7.2000     3.017134   1.736990   
volatile acidity       0.529478   0.5200   0.5000     0.033500   0.183031   
citric acid            0.272333   0.2600   0.0000     0.038235   0.195537   
residual sugar         2.523400   2.2000   2.0000     1.828752   1.352314   
chlorides              0.088124   0.0790   0.0800     0.002438   0.049377   
free sulfur dioxide   15.893304  14.0000   6.0000   109.145456  10.447270   
total sulfur dioxide  46.825975  38.0000  28.0000  1116.157653  33.408946   
density                0.996709   0.9967   0.9968     0.000003   0.001869   
pH                     3.309787   3.3100   3.3000     0.024036   0.155036   
sulphates              0.658705   0.6200   0.5400     0.029127   0.170667   
alcohol               10.432315  10.2000   9.5000     1.170866   1.082065   
quality                5.623252   6.0000   5.0000     0.678281   0.823578   

                          range  percentile_25  percentile_50  percentile_75  \
fixed acidity          11.30000         7.1000         7.9000        9.20000   
volatile acidity        1.46000         0.3900         0.5200        0.64000   
citric acid             1.00000         0.0900         0.2600        0.43000   
residual sugar         14.60000         1.9000         2.2000        2.60000   
chlorides               0.59900         0.0700         0.0790        0.09100   
free sulfur dioxide    71.00000         7.0000        14.0000       21.00000   
total sulfur dioxide  283.00000        22.0000        38.0000       63.00000   
density                 0.01362         0.9956         0.9967        0.99782   
pH                      1.27000         3.2100         3.3100        3.40000   
sulphates               1.67000         0.5500         0.6200        0.73000   
alcohol                 6.50000         9.5000        10.2000       11.10000   
quality                 5.00000         5.0000         6.0000        6.00000   

                           iqr  
fixed acidity          2.10000  
volatile acidity       0.25000  
citric acid            0.34000  
residual sugar         0.70000  
chlorides              0.02100  
free sulfur dioxide   14.00000  
total sulfur dioxide  41.00000  
density                0.00222  
pH                     0.19000  
sulphates              0.18000  
alcohol                1.60000  
quality                1.00000

Tư kết quả này, ta có thể sử dụng chúng để thực các công việc tiếp theo (như trực quan hóa dữ liệu, xử lý ngoại lai, ...)

## 5. Trực quan hóa dữ liệu